056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,159 @@
+This call fails because the quorum is not formed, so in some cases this results in the cluster being stuck with some pods 
+in a pending state.
+
+## Motivation
+
+When running in the KRaft mode the controller pods need to be rolled if manually annotated, configuration changes occur or pod is unready. At the moment the 
+existing logic is blocking the Zookeeper to KRaft migration that is being proposed in [PR #90](https://github.com/strimzi/proposals/pull/90) as well as the full implementation of KRaft liveness and readiness checks as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md).
+
+## Proposal
+
+The KafkaRoller behaviour should be unchanged when operating against a ZooKeeper based cluster.
+
+The behaviour when operating against a KRaft cluster is described below.
+
+### Order to restart nodes
+Restart in order:
+1. Unready controller/combined nodes
+2. Ready controller/combined nodes in follower state
+3. Active controller (applies to both pure controller and combined node)
+4. Unready broker-only nodes
+5. Ready broker-only nodes
+
+
+### Triggers
+The following are some of the triggers that would restart a KRaft controller or combined node:
+- Pod or its StrimziPodSet is annotated for manual rolling update
+- Pod is unready 
+- Controller's configuration has changed
+
+The triggers for broker remain the same as Zookeeper mode.
+
+### Restart conditions
+The restart conditions that the KafkaRoller considers in different modes are described below.
+
+#### The new quorum check
+The restart conditions include a new check for controllers to verify that restarting the Kafka node does not affect the quorum health.
+The proposed check is:
+- Create admin client connection to the brokers and call `describeMetadataQuorum` API
+- If failed to connect to the brokers, return `UnforceableProblem` which currently results in delay and retry for the pod until the maximum attempts is reached.
+- From the quorum info returned from the admin API, read `lastCaughtUpTimestamp` of each controller. `lastCaughtUpTimestamp` is the last millisecond timestamp at which a replica controller was known to be caught up with the quorum leader.
+- Check the quorum leader id using the quorum info and identify the `lastCaughtUpTimestamp` of the quorum leader
+- Retrieve value of the Kafka property `controller.quorum.fetch.timeout.ms` from the desired configurations specified in the Kafka CR. If this property does not exist in the desired configurations, then use the hard-coded default value for it which is `2000`. The reason for this is explained further in the **NOTE** below.
+- Mark a Kafka node as caught up if `leaderLastCaughtUpTimestamp - replicaLastCaughtUpTimestamp < controllerQuorumFetchTimeoutMs`. This will exclude the controller KafkaRoller is currently considering to restart, because we are trying to check whether the quorum would stay healthy when this controller is restarted.
```

Comment:
```suggestion
- Mark a KRaft controller node as caught up if `leaderLastCaughtUpTimestamp - replicaLastCaughtUpTimestamp < controllerQuorumFetchTimeoutMs`. This will exclude the controller KafkaRoller is currently considering to restart, because we are trying to check whether the quorum would stay healthy when this controller is restarted.
```